File: src/main/java/com/circleescape/server/model/Game.java
package com.circleescape.server.model;

import lombok.Getter;
import lombok.extern.log4j.Log4j2;

@Log4j2
public class Game {
	
	private double speedFactor;
	private double radius;
	
	private PolarCoordinates escapee;
	private double guard;
	
	private GuardStrategy guardStrategy = new GreedyGuard();
	
	TurnResult state;
	
	@Getter
	private boolean hasbeeninitialized;
	
	private void logf(String format, Object... args) {
		log.info(String.format(format, args));
	}
	
	//TODO learn how to initialize this bean in constructor
	public void initialize(GameParameters gp) {
		this.speedFactor = gp.getSpeedFactor();
		this.radius = gp.getRadius();
		this.escapee = new PolarCoordinates(0, MathUtils.TAU/4);
		this.guard = MathUtils.TAU*3/4;
		this.state = TurnResult.ONGOING;
		logf("this is game");
		logf(this.toString());
		logf("escapee null? %b",this.escapee==null);
		hasbeeninitialized=true;
	}
	
	public GameParameters getParameters() {
		return new GameParameters(radius, speedFactor);
	}
	
	public GameState getGameState() {
		logf("game state");
		logf("hasbeeninitialized: %b",hasbeeninitialized);
		logf("this is game");
		logf(this.toString());
		logf("escapee null? %b",this.escapee==null);

		Cat cat = new Cat(guard);
		GameState gs = new GameState(speedFactor, radius, escapee, cat, state);

		return gs;
	}
	
	public Positions getPositions() {
		return new Positions(escapee, new PolarCoordinates(radius, guard));
	}
	
	public Result moveEscapee(Diff d) {
		return null;
	}

	public TurnResponse predictGameState(PolarCoordinates newDuck) {
		
		newDuck = newDuck.normalize(radius);
		
		double guard = guardStrategy.updateCat(getParameters(), this.guard, escapee, newDuck);
		
		TurnResult state = duckAtTheEdge(newDuck) ? catAtDuck(guard, newDuck) ? TurnResult.LOOSE
		                                                                      : TurnResult.WIN
		                                          : TurnResult.ONGOING;
		
		return new TurnResponse(newDuck, guard, state);
	}

	
	public TurnResponse updateEscapeePosition(PolarCoordinates newEscapeePosition) {
		TurnResponse tr = predictGameState(newEscapeePosition);
		this.escapee = tr.getDuck();
		this.guard = tr.getCat().getTheta();
		this.state = tr.getState();
		return tr;
	}
	
	private boolean duckAtTheEdge(PolarCoordinates duck) {
		return Math.abs(radius - duck.getR()) < MathUtils.ϵ;
	}
	
	private boolean catAtDuck(double cat, PolarCoordinates duck) {
		return Math.abs(cat - duck.getTheta()) < MathUtils.ϵ;
	}


}

File: src/main/kotlin/com/circleescape/server/model/Game.kt
package com.circleescape.server.model

import com.circleescape.server.api.game.model.GameStatus
import org.slf4j.LoggerFactory
import kotlin.math.abs

class Game constructor(val speedFactor: Double = 4.0,
           val radius: Double = 1.0,
           var escapee: PolarCoordinates,
           var guard: Double = 0.0,
           var state: GameStatus
) {

    private val log = LoggerFactory.getLogger(this.javaClass)

    private val guardStrategy: GuardStrategy = GreedyGuard()

    constructor(gp: GameParameters) : this(
        gp.speedFactor,
        gp.radius,
        PolarCoordinates(0.0, MathUtils.TAU / 4),
        MathUtils.TAU * 3 / 4,
        GameStatus.ONGOING
    )

    val parameters: GameParameters
        get() = GameParameters(radius, speedFactor)

    val gameState: GameState
        get() {
            val cat = Cat(guard)
            return GameState(speedFactor, radius, escapee, cat, state)
        }

    val positions: Positions
        get() = Positions(escapee, guard)

    fun moveEscapee(d: Diff?): Result? {
        return null
    }


}